project_euler/problem_54/sol1.py

```diff
@@ -0,0 +1,366 @@
+
+    # This function is not part of the problem, I did it just for fun
+    def hand_name(self) -> str:
+        """
+        Return the name of the hand in the following format:
+        'hand name, high card'
+
+        Here are some examples:
+        >>> PokerHand("KS AS TS QS JS").hand_name()
+        'Royal flush'
+
+        >>> PokerHand("2D 6D 3D 4D 5D").hand_name()
+        'Straight flush, Six-high'
+
+        >>> PokerHand("JC 6H JS JD JH").hand_name()
+        'Four of a kind, Jacks'
+
+        >>> PokerHand("3D 2H 3H 2C 2D").hand_name()
+        'Full house, Twos over Threes'
+
+        >>> PokerHand("2H 4D 3C AS 5S").hand_name()  # Low ace
+        'Straight, Five-high'
+
+        Source: https://en.wikipedia.org/wiki/List_of_poker_hands
+        """
+        name = PokerHand._HAND_NAME[self._hand_type - 14]
+        high = PokerHand._CARD_NAME[self._high_card]
+        pair1 = PokerHand._CARD_NAME[self._first_pair]
+        pair2 = PokerHand._CARD_NAME[self._second_pair]
+        if self._hand_type in [22, 19, 18]:
+            return name + f", {high}-high"
+        elif self._hand_type in [21, 17, 15]:
+            return name + f", {pair1}s"
+        elif self._hand_type in [20, 16]:
+            join = "over" if self._hand_type == 20 else "and"
+            return name + f", {pair1}s {join} {pair2}s"
+        elif self._hand_type == 23:
+            return name
+        else:
+            return name + f", {high}"
+
+    def _compare_cards(self, other: "PokerHand") -> str:
+        # Comparing in reverse order as the list is sorted
+        # 'i' starts at 4 and decreases by 1 up to 0
+        # Remember, end value is not included in the range function
+        for i in range(4, -1, -1):
+            if self._cards[i][0] != other._cards[i][0]:
+                return "Win" if self._cards[i][0] > other._cards[i][0] else "Loss"
+        return "Tie"
+
+    def _get_hand_type(self) -> int:
+        # Number representing the type of hand internally:
+        # 23: Royal flush
+        # 22: Straight flush
+        # 21: Four of a kind
+        # 20: Full house
+        # 19: Flush
+        # 18: Straight
+        # 17: Three of a kind
+        # 16: Two pairs
+        # 15: One pair
+        # 14: High card
+        if self._is_flush():
+            if self._is_five_high_straight() or self._is_straight():
+                if sum(self._card_values) == 60:
+                    return 23
+                else:
+                    return 22
+            return 19
+        elif self._is_five_high_straight() or self._is_straight():
+            return 18
+        return 14 + self._is_same_kind()
+
+    def _get_high_card(self) -> int:
+        return self._cards[-1][0]
+
+    def _card_value_list(self) -> list:
+        return list(map(itemgetter(0), self._cards))
+
+    def _is_flush(self) -> bool:
+        suit = self._cards[0][-1]
+        return all(map(lambda card: suit in card, self._cards))
```

Comment:
```suggestion
        return all(suit in card for card in self._cards)
```
As discussed in CONTRIBUTING.md.

Reply:
Ah, missed that. Thanks for pointing it out and yes this looks more readable. I will make the following changes in my next commit:
- Remove `map` and use list comprehension
- Sort imports using `isort`
- Sort in reverse? (still in discussion)